src/storage/txn/txn_status_cache.rs

```diff
@@ -0,0 +1,962 @@
+            slots: (0..slots)
+                .map(|_| {
+                    let cache = LruCache::new(
+                        allowed_capacity_per_slot,
+                        0,
+                        lru::CountTracker::default(),
+                        TxnStatusCacheEvictPolicy::new(
+                            required_keep_time,
+                            simulated_system_time.clone(),
+                        ),
+                    );
+                    let allocated_capacity = cache.internal_allocated_capacity();
+                    initial_allocated_capacity_total += allocated_capacity;
+                    Mutex::new(cache).into()
+                })
+                .collect(),
+            is_enabled: true,
+        };
+        SCHED_TXN_STATUS_CACHE_SIZE
+            .allocated
+            .set(initial_allocated_capacity_total as i64);
+        res
+    }
+
+    pub fn new(capacity: usize) -> Self {
+        Self::with_slots_and_time_limit(
+            TXN_STATUS_CACHE_SLOTS,
+            CACHE_ITEMS_REQUIRED_KEEP_TIME,
+            capacity,
+        )
+    }
+
+    #[cfg(test)]
+    pub fn new_for_test() -> Self {
+        // 1M capacity should be enough for tests.
+        Self::with_slots_and_time_limit(16, CACHE_ITEMS_REQUIRED_KEEP_TIME, 1 << 20)
+    }
+
+    pub fn with_slots_and_time_limit(
+        slots: usize,
+        required_keep_time: Duration,
+        capacity: usize,
+    ) -> Self {
+        Self::new_impl(slots, required_keep_time, capacity, None)
+    }
+
+    /// Create a `TxnStatusCache` instance for test purpose, with simulating
+    /// system time enabled. This helps when testing functionalities that are
+    /// related to system time.
+    ///
+    /// An `AtomicU64` will be returned. Store timestamps
+    /// in milliseconds in it to control the time.
+    #[cfg(test)]
+    fn with_simulated_system_time(
+        slots: usize,
+        requried_keep_time: Duration,
+        capacity: usize,
+    ) -> (Self, Arc<AtomicU64>) {
+        let system_time = Arc::new(AtomicU64::new(0));
+        let res = Self::new_impl(
+            slots,
+            requried_keep_time,
+            capacity,
+            Some(system_time.clone()),
+        );
+        (res, system_time)
+    }
+
+    fn slot_index(&self, start_ts: TimeStamp) -> usize {
+        fxhash::hash(&start_ts) % self.slots.len()
+    }
+
+    /// Insert a transaction status into the cache. The current system should
+    /// be passed from outside to avoid getting system time repeatedly when
+    /// multiple items is being inserted.
+    ///
+    /// If the transaction's information is already in the cache, it will
+    /// **NOT** be promoted to the most-recent place of the internal LRU.
+    pub fn insert(&self, start_ts: TimeStamp, commit_ts: TimeStamp, now: SystemTime) {
+        if !self.is_enabled {
+            return;
+        }
+
+        let insert_time = now.duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
```

Comment:
How much is its overhead compared with the `now_coarse` like functions?

Reply:
This line is actually to calculate the integer timestamp from the `SystemTime` type, and it didn't acquire any time from system. The variable `now` is passed from outside the function.
As for the time cost, I have such benchmark results in my environment: (the last one `bench_system_time_since_unix_epoch` is the calculation in the line you commented)

```
bench_instant_now                               ... bench:          19 ns/iter (+/- 1)
bench_instant_now_coarse                        ... bench:           9 ns/iter (+/- 0)
bench_system_time_now                           ... bench:          19 ns/iter (+/- 0)
bench_system_time_since_unix_epoch              ... bench:           5 ns/iter (+/- 0)
```

Reply:
Got, it looks good not to acuiqre system time.

Reply:
Sorry but you may need to notice that the `should_evict` function called by the `insert` function  still need to acquire the system time, and I failed to avoid it without making the code too ugly...